main.c

```diff
@@ -137,7 +137,18 @@ static void linux_boot(const struct boot_params boot_params)
     printf("  pref_address: 0x%llx\n", l_params->pref_address);
     printf("\n");
 
-    char *linux_header_end = (char *)(boot_params.kernel_addr + 0x201);
+    // If the kernel is not relocatable, we try to load Linux at the aaddress it wants to have.
+    uintptr_t actual_kernel_load_addr;
+    if (l_params->relocatable_kernel) {
+        actual_kernel_load_addr = boot_params.kernel_addr;
+        printf("Kernel is relocatable. We use the load address from the VMM\n");
+    } else {
+        actual_kernel_load_addr = l_params->pref_address;
+        printf("Kernel is not relocatable. We try the preferred load address from the kernel\n");
+    }
```

Comment:
I don't think this distinction makes sense. The VMM should not force the kernel load address in any case. It's up to the loader to find a suitable place. And the "preferred address" is always the preferred address.

The VMM actually placed the kernel image at `kernel_addr`, and communicates that address via the `fw_cfg` entry. A relocatable kernel will relocate itself on boot if possible, while a non-relocatable kernel would die. Instead, the loader should move the kernel from the original `kernel_addr` to the preferred address immediately, making life easier for all bzImages. This includes checking if the current address is the preferred address (no relocation required) and if the preferred address conflicts with the kernel, initrd, or cmdline placed by the VMM (from `fw_cfg`).

The relocation should probably use `memmove` as the source and destination of the kernel might overlap.